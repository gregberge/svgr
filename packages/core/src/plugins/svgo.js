import SVGO from 'svgo'
import cosmiconfig from 'cosmiconfig'
import mergeDeep from 'merge-deep'

const explorer = cosmiconfig('svgo', {
  searchPlaces: [
    'package.json',
    `.svgorc`,
    `.svgorc.json`,
    `.svgorc.yaml`,
    `.svgorc.yml`,
    `svgo.config.js`,
    '.svgo.yml',
  ],
  transform: result => result && result.config,
})

function getBaseSvgoConfig(config) {
  const baseSvgoConfig = { plugins: [] }
  if (config.icon) baseSvgoConfig.plugins.push({ removeViewBox: false })
  return baseSvgoConfig
}

export default async (code, config = {}, state = {}) => {
  if (!config.svgo) return code
  const filePath = state.filePath || process.cwd()
  const svgoRcConfig = await explorer.search(filePath)
  const svgo = new SVGO(
    mergeDeep(getBaseSvgoConfig(config), svgoRcConfig, config.svgoConfig),
  )
  const info = state.filePath
    ? { input: 'file', path: state.filePath }
    : { input: 'string' }
  const { data } = await svgo.optimize(code, info)
  return data
}
