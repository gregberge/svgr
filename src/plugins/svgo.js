import SVGO from 'svgo'
import cosmiconfig from 'cosmiconfig'
import merge from 'lodash/merge'

const explorer = cosmiconfig('svgo', {
  searchPlaces: [
    'package.json',
    `.svgorc`,
    `.svgorc.json`,
    `.svgorc.yaml`,
    `.svgorc.yml`,
    `svgo.config.js`,
    '.svgo.yml',
  ],
  transform: result => result && result.config,
})

export default async (code, opts = {}, state = {}) => {
  const filePath = state.filePath || process.cwd()
  const config = await explorer.search(filePath)
  const svgo = new SVGO(merge({}, config, opts))
  const { data } = await svgo.optimize(code)
  return data
}
