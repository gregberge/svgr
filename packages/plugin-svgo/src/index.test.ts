import * as path from 'path'
import svgo from '.'

const state = { componentName: 'SvgComponent' }

const baseSvg = `<?xml version="1.0" encoding="UTF-8"?>
<svg width="88px" height="88px" viewBox="0 0 88 88" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
  <!-- Generator: Sketch 46.2 (44496) - http://www.bohemiancoding.com/sketch -->
  <title>Dismiss</title>
  <desc>Created with Sketch.</desc>
  <defs></defs>
  <style>.shape{fill: red}</style>
  <g id="Blocks" class="blocks" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd" stroke-linecap="square">
      <g id="Dismiss" stroke="#063855" stroke-width="2">
          <path d="M51,37 L37,51" id="Shape" class="shape"></path>
          <path d="M51,51 L37,37" id="Shape" class="shape"></path>
          <style>
            #Shape {}
          </style>
      </g>
  </g>
</svg>`

describe('svgo', () => {
  it('optimizes svg', () => {
    const result = svgo(baseSvg, { svgo: true, runtimeConfig: true }, state)
    expect(result).toMatchSnapshot()
  })

  it('supports `config.svgoConfig`', () => {
    const result = svgo(
      baseSvg,
      {
        svgo: true,
        runtimeConfig: true,
        svgoConfig: {
          plugins: [
            {
              name: 'preset-default',
              params: {
                overrides: {
                  removeDesc: false,
                },
              },
            },
          ],
        },
      },
      state,
    )
    expect(result).toMatchSnapshot()
  })

  it('throws error on invalid svg input', () => {
    const errorSvg = `<?xml version="1.0" encoding="UTF-8"?>
  <svg width="88px" height="88px" viewBox="0 0 88 88" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <path d=M51,37 L37,51" id="Shape" class="shape"></path>
  </svg>`

    expect(() =>
      svgo(
        errorSvg,
        { svgo: true, runtimeConfig: true },
        { ...state, filePath: path.join(__dirname, '../__fixtures__/svgo') },
      ),
    ).toThrowError()
  })

  it('uses `state.filePath` to detect configuration', () => {
    const result = svgo(
      baseSvg,
      { svgo: true, runtimeConfig: true },
      { ...state, filePath: path.join(__dirname, '../__fixtures__/svgo') },
    )

    expect(result).toMatchSnapshot()
  })

  it('does not load runtime configuration with `runtimeConfig: false`', () => {
    const result = svgo(
      baseSvg,
      { svgo: true, runtimeConfig: false },
      { ...state, filePath: path.join(__dirname, '../__fixtures__/svgo') },
    )
    expect(result).toMatchSnapshot()
  })

  it('does not remove viewBox with `icon` option', () => {
    const result = svgo(baseSvg, { svgo: true, icon: true }, state)
    expect(result).toMatchSnapshot()
  })

  it('does not remove viewBox with when `dimensions` is false', () => {
    const result = svgo(baseSvg, { svgo: true, dimensions: false }, state)
    expect(result).toMatchSnapshot()
  })

  it('does remove style when `native` is true', () => {
    const result = svgo(baseSvg, { svgo: true, native: true }, state)
    expect(result).toMatchSnapshot()
  })
})
