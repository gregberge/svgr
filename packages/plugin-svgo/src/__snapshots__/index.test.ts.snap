// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`svgo does not load runtime configuration with \`runtimeConfig: false\` 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="88" height="88"><style>.svgo__shape{fill:red}</style><g id="svgo__Blocks" fill="none" fill-rule="evenodd" stroke="none" stroke-linecap="square" stroke-width="1" class="svgo__blocks"><g id="svgo__Dismiss" stroke="#063855" stroke-width="2"><path id="svgo__Shape" d="M51 37 37 51" class="svgo__shape"/><path d="M51 51 37 37" class="svgo__shape"/></g></g></svg>"`;

exports[`svgo does not remove viewBox with \`icon\` option 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="88" height="88" viewBox="0 0 88 88"><style>.prefix__shape{fill:red}</style><g id="prefix__Blocks" fill="none" fill-rule="evenodd" stroke="none" stroke-linecap="square" stroke-width="1" class="prefix__blocks"><g id="prefix__Dismiss" stroke="#063855" stroke-width="2"><path id="prefix__Shape" d="M51 37 37 51" class="prefix__shape"/><path d="M51 51 37 37" class="prefix__shape"/></g></g></svg>"`;

exports[`svgo does not remove viewBox with when \`dimensions\` is false 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="88" height="88" viewBox="0 0 88 88"><style>.prefix__shape{fill:red}</style><g id="prefix__Blocks" fill="none" fill-rule="evenodd" stroke="none" stroke-linecap="square" stroke-width="1" class="prefix__blocks"><g id="prefix__Dismiss" stroke="#063855" stroke-width="2"><path id="prefix__Shape" d="M51 37 37 51" class="prefix__shape"/><path d="M51 51 37 37" class="prefix__shape"/></g></g></svg>"`;

exports[`svgo does remove style when \`native\` is true 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="88" height="88"><g fill="none" fill-rule="evenodd" stroke="#063855" stroke-linecap="square" stroke-width="2" class="prefix__blocks"><path d="M51 37 37 51M51 51 37 37" style="fill:red"/></g></svg>"`;

exports[`svgo optimizes svg 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="88" height="88"><style>.prefix__shape{fill:red}</style><g id="prefix__Blocks" fill="none" fill-rule="evenodd" stroke="none" stroke-linecap="square" stroke-width="1" class="prefix__blocks"><g id="prefix__Dismiss" stroke="#063855" stroke-width="2"><path id="prefix__Shape" d="M51 37 37 51" class="prefix__shape"/><path d="M51 51 37 37" class="prefix__shape"/></g></g></svg>"`;

exports[`svgo supports \`config.svgoConfig\` 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="88" height="88"><desc>Created with Sketch.</desc><style>.shape{fill:red}</style><g id="Blocks" fill="none" fill-rule="evenodd" stroke="none" stroke-linecap="square" stroke-width="1" class="blocks"><g id="Dismiss" stroke="#063855" stroke-width="2"><path id="Shape" d="M51 37 37 51" class="shape"/><path d="M51 51 37 37" class="shape"/></g></g></svg>"`;

exports[`svgo uses \`state.filePath\` to detect configuration 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="88" height="88"><style>.svgo__shape{fill:red}</style><g id="svgo__Blocks" fill="none" fill-rule="evenodd" stroke="none" stroke-linecap="square" stroke-width="1" class="svgo__blocks"><g id="svgo__Dismiss" stroke="#063855" stroke-width="2"><path id="svgo__Shape" d="M51 37 37 51" class="svgo__shape"/><path d="M51 51 37 37" class="svgo__shape"/></g></g></svg>"`;
